src/viam/sdk/services/mlmodel/private/proto.cpp

```diff
@@ -0,0 +1,333 @@
+// Copyright 2023 Viam Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <viam/sdk/services/mlmodel/private/proto.hpp>
+
+#include <stack>
+
+#include <boost/variant/get.hpp>
+
+namespace viam {
+namespace sdk {
+namespace mlmodel_details {
+
+namespace {
+
+template <typename T>
+::grpc::Status pb_value_to_tensor_t(const MLModelService::tensor_info& tensor_info,
```

Comment:
Note that these methods (`pb_value_to_tensor_t` and `tensor_to_pb_`) do not yet have good test coverage. I wouldn't spend a ton of time worrying about correctness for them just yet, as I'd consider them very likely to have a few issues.